File: Mathematic/Optimisation/Python/dichotomie.py

from math import exp,cos,sin

f = lambda x: exp(x)+x**3;
g = lambda y: cos(y)+sin(y);

def dichotomie(f, a, b, epsilon, Niter):
    m=(a+b)/2;
    i=0;
        
    while(i<Niter):
        if(epsilon>abs(f(m))):
            print(" L'algorithme de dichotomie donne ", m, "et i vaut", i, ".\n");
            break;
        if(f(m)<0):
            a=m;
        else:
            b=m;
        m=(a+b)/2
        i=i+1;
    if(i==Niter):
        print("L'algorithme n'a pas convergé au bout de ", Niter,"itérations \n");


dichotomie(f, -10,10,1e-15,1000);
dichotomie(g, -10,10,1e-15,1000);

'''
global Compte;
global Non_coverger;
def dicho(f, a, b, epsilon, Niter):
    Compte=0;
    Non_coverger=0;
    m=(a+b)/2;
    i=0;
    while(i<Niter):
        if(epsilon>abs(f(m))):
            Compte=Compte+1;
            dicho(f, m+epsilon, b, epsilon, Niter);
            dicho(f, a, m-epsilon, epsilon, Niter);
        if(f(m)<0):
            a=m;
        else:
            b=m;
        m=(a+b)/2
        i=i+1;
    if(i==Niter):
        Non_coverger=Non_coverger+1;


dicho(f, -1, 1, 1e-15,100);
print(" Compte ", Compte," Non_converger ", Non_coverger);
'''

File: Mathematic/Optimisation/C++/dichotomie.cpp

#include <iostream>
#include <cmath>
using namespace std;

auto f = [](long double x) { return exp(x)+x*x*x; };

float absolue(long double x){
    if(x<0){
        return -x;
    }
    return x;
}

void dichotomie(long double a, long double b,long double epsilon, int Niter){
    float m=0;
    m=(a+b)/2;
    int i=0;
        
    while(i<Niter){
        if(epsilon>absolue(f(m))){
            cout << " L'algorithme de dichotomie donne "<< m<< " et i vaut "<< i<<".\n"<<endl;
            break;
        }
        if(f(m)<0){
            a=m;
        }
        else{
            b=m;
        }
        m=(a+b)/2;
        i=i+1;
    }
    if(i==Niter){
        cout << "L'algorithme n'a pas convergé au bout de " << Niter << "itérations \n"<<endl;
    }
}

int main(int argc, char** argv){

    dichotomie(-10,10,1e-5,100);

    return 0;
}

